src/test/java/com/diffmin/AppTest.java

```diff
@@ -73,22 +72,26 @@
          * @return instance of {@link TestResources}
          */
         public static TestResources fromTestDirectory(File testDir) {
-            Path path = testDir.toPath();
             String parent = testDir.getName();
-            Stream<Path> paths = Arrays.stream(path.toFile().listFiles()).map(File::toPath);
-            List<Path> pathList = paths.collect(Collectors.toList());
-            Map<String, Path> testPathMap = new HashMap<>();
-            for (Path testPath : pathList) {
-                if (testPath.getFileName().toString().startsWith(AppTest.PREV_PREFIX)) {
-                    testPathMap.put(AppTest.PREV_PREFIX, testPath);
-                } else if (testPath.getFileName().toString().startsWith(AppTest.NEW_PREFIX)) {
-                    testPathMap.put(AppTest.NEW_PREFIX, testPath);
-                }
-            }
-            return new TestResources(
-                    testPathMap.get(AppTest.PREV_PREFIX),
-                    testPathMap.get(AppTest.NEW_PREFIX),
-                    parent);
+            Path prevPath = getFilepathByPrefix(testDir, PREV_PREFIX);
+            Path newPath = getFilepathByPrefix(testDir, NEW_PREFIX);
+            return new TestResources(prevPath, newPath, parent);
+        }
+
+        /**
+         * Returns test resource with the matching prefix.
+         *
+         * @param dir Directory where the test resources are located
+         * @param prefix Prefix of the test resource
+         * @return {@link Path} to the test resource
+         */
+        private static Path getFilepathByPrefix(File dir, String prefix) {
+            return Arrays.stream(dir.listFiles())
+                    .filter(f -> f.getName().startsWith(prefix))
+                    .findFirst()
+                    .map(File::toPath)
+                    .orElseThrow(
+                            () -> new IllegalArgumentException("Filename has an invalid prefix"));
```

Comment:
Just need an advice here. Do you recommend that I should add "Valid prefixes are `PREV` and `NEW`."? I felt some thing like this belongs more in a `Contributing.md` file and I wanted to keep the error message concise.

Reply:
I would:

1. Probably throw a `FileNotFoundException` here. It's not really a matter of an illegal argument, we just couldn't find the file we were looking for. Ps. don't remember what exception I used in the live coding session, but I didn't think twice.
2. Have something like `String.format("Expected file with prefix '%s' in directory '%s', prefix, dir)` in the message.
3. Yes, the valid prefixes should go into some contributing documentation, but that's not for this PR.

Reply:
Also note that we could be even stricter here, and require not only that we find a file with the given prefix, but that we find _exactly one_ file with that prefix. I think checking that we find one at all is sufficient, though.

Reply:
> Probably throw a FileNotFoundException here

I thought to do that but `FileNotFoundException` is a checked exception so we will have to add it to the signature or use try-catch everywhere. Should I create `FileNotFoundRuntimeExcepting` by inheriting from `RuntimeException`?

Reply:
Just declare `throws FileNotFoundException` to the signature of affected methods. Top level tests can throw any exception they'd like (i.e. declare to throw any exception), JUnit takes care of it.

Reply:
> Should I create FileNotFoundRuntimeException by inheriting from RuntimeException?

This would circumvent the point of checked exceptions, the point being that they _force_ you to declare in the API of the method what exceptions are thrown, which is good! Unfortunately, it poses problems with streams, and so isn't always practical (not allowed to throw checked exceptions in streams).

Reply:
Don't you think it would be better to use an unchecked exception here? I am referring to this [StackOverflow post](https://stackoverflow.com/a/19061110) to decide that. As this exception is very much preventable, we should use a `RuntimeException`.

Reply:
>  the point being that they force you to declare in the API of the method what exceptions are thrown

Yes, that is true. I like this advantage that checked exceptions have.

Reply:
```java
return getArgumentSourceStream(
                    PURE_INSERT_PATCHES.toFile(), TestResources::fromTestDirectory);
```

This line throws an error despite adding Exception to its signature.

Reply:
> As this exception is very much preventable

If we view the test code _and_ resources as one entity, then it could be considered preventable. If we don't, then from the code side of the table, it's unpreventable. In both cases, it's of course unrecoverable.

Anyway, there's no one interpretation of when to use checked exceptions and unchecked exceptions (the extreme opinions being _only use checked_ and _only use unchecked_).

> This line throws an error despite adding Exception to its signature.

There you have another limitation of checked exceptions: they are part of the API, and you're not allowed to make breaking changes to the API of an overridden method. So that settles it, use a runtime exception instead :)

Reply:
Personally, as I'm a big fan of using streams, I typically steer clear of checked exceptions. In combination, they are a headache.

Reply:
[Another view](https://www.baeldung.com/java-checked-unchecked-exceptions) is that checked exceptions should be used for stuff outside of the program's control (such as external resources), and unchecked exceptions for logical errors (i.e. bugs). Again, there's no consensus here.

Reply:
> In both cases, it's of course unrecoverable.

It is unrecoverable because we cannot rename the file at runtime? I mean we can if we want to but we don't want to, of course.

> there's no one interpretation of when to use checked exceptions and unchecked exceptions

Yep, Even I haven't reached a concrete conclusion yet.

> Unfortunately, it poses problems with streams, and so isn't always practical (not allowed to throw checked exceptions in streams).

What problems exactly where you taking about? I thought you meant in the context of `public Stream<? extends Arguments> provideArguments` but I think you were talking about a more general case.

> So that settles it, use a runtime exception instead :)

Cool. I used a `IllegalFormatException` initially but its constructor is package private so I am unable to access it. I will create my own custom exception and will add a proper error message.

Reply:
> It is unrecoverable because we cannot rename the file at runtime? I mean we can if we want to but we don't want to, of course.

Which file should we rename? What if the file name isn't just misspelled, but instead the file just isn't there? There's no general solution to not finding a file with the expected prefix.

> What problems exactly where you taking about? I thought you meant in the context of public Stream<? extends Arguments> provideArguments but I think you were talking about a more general case.

No, that's not what I was talking about. You're not allowed to throw a checked exception in a stream, or call a method that does without converting to a runtime exception. For example, this is illegal:

```java
public Stream<Integer> method() throws Exception {
  return Stream.of(1, 2, 3).map(i -> {
      if (i < 2) {
          throw new Exception();
      }
      return i;
  });
}
```

because even though the method where the stream is declared says that it throws some checked exception, the method where the stream is later evaluated perhaps doesn't, and so it breaks the whole thing. And streams can't themselves be declared to throw stuff (as that would require a _type_ to be declared as throwing something, which isn't a concept that exists in Java).

> Cool. I used a IllegalFormatException initially but its constructor is package private so I am unable to access it. I will create my own custom exception and will add a proper error message.

That's not what `IllegalFormatException` is for, [read its docs here](https://docs.oracle.com/javase/8/docs/api/java/util/IllegalFormatException.html) :). The closest I can think of to this particular thing is `IllegalStateException`, although that's technically about an illegal state in the application, but at least the name makes sense.

This is way overthinking it though, you could use `RuntimeException` and it'd be fine. It doesn't matter.

Reply:
> because even though the method where the stream is declared says that it throws some checked exception, the method where the stream is later evaluated perhaps doesn't, and so it breaks the whole thing

Sounds good. Thanks for the example.

> That's not what IllegalFormatException

I took `Format` in its literal sense. 😅 

> you could use RuntimeException

I created a custom exception (called `FileNotFoundAtRuntimeException`) which inherits from this same class. I just felt that the class name would be more intuitive. Let me know if that's alright.

Reply:
> I created a custom exception (called FileNotFoundAtRuntimeException) which inherits from this same class. I just felt that the class name would be more intuitive. Let me know if that's alright.

There's nothing wrong with doing that. That being said, I personally wouldn't have bothered as it's used in a test, and adds code to the codebase. I think a `RuntimeException` with a clear message is fine here. Perhaps a generically named `IllegalTestStateException` would be more reusable for other things in the test suite.

Anyway, custom exceptions can be a _great_ way to communicate problems to consumers of your code, but in a test suite where everyone who gets exposed to the exception is a developer of the application, it's less important in my opinion.

Reply:
> Perhaps a generically named IllegalTestStateException would be more reusable for other things in the test suite.

Oh, okay. That sounds like a good idea. I will write one if we ever find some more problems.

> I personally wouldn't have bothered as it's used in a test, and adds code to the codebase. I think a RuntimeException

Yeah, makes sense 😅 I should create a mental separation between tests and main so that I avoid adding syntactical sugar to the tests as well.

Reply:
> Yeah, makes sense sweat_smile I should create a mental separation between tests and main so that I avoid adding syntactical sugar to the tests as well.

It's not easy to balance. One shouldn't get _sloppy_ with test code; it's not less important than production code, but the prioritization between the two is different due to the different use cases.

I will have to defer a full review of this PR until some other day, need to clock out now :)